server/src/main/resources/db/migration/V3_0__Scope_Permissions_Grant_ID.sql

```diff
@@ -0,0 +1,10 @@
+ALTER TABLE scope_permissions ADD COLUMN grant_id text;
+
+CREATE INDEX IF NOT EXISTS scope_permissions_grant_id_idx ON scope_permissions(grant_id);
+
+-- Remove old index that enforces unique on scope_address/grantee_address/granter_address
+DROP INDEX IF EXISTS scope_permissions_scope_address_grantee_address_granter_address;
+
+-- LMAO BEST NAME EVER.  Ensures that there can be duplicate grants from granters to grantees on a scope address, as
+-- long as the duplicate specifies a null or newly-unique grant id
+CREATE UNIQUE INDEX IF NOT EXISTS scope_permissions_scope_addr_grantee_addr_granter_addr_grant_id ON scope_permissions (scope_address, grantee_address, granter_address, grant_id);
```

Comment:
does a combo of the 3 addresses and a null grant_id enforce uniqueness with the another exactly the same (with the null)? I feel like I remember running into an issue with this in the past where null != null... that was mysql though, not sure if postgres handles that a bit differently or not...

Reply:
Probably doesn't matter much since we do the check in code before inserting, but I am curious 🤔

Reply:
You're absolutely right.  I hacked together a test, and the database does appear to accept multiple null records.  The `new` function 100% prevents this, but only if we use it... I wonder if there's a postgres-ish way to avoid this behavior and regard null as a combination of the unique key

Reply:
Thanks Ravi, I just got this added and it totally works!

Reply:
oh that's a good tip, didn't know you can qualify indexes like that, thanks @rpatel-figure!